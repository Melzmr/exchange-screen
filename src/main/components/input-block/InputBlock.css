.input_block_inner_container {
    display: flex;
    height: 100%;
}

.input_block_col_left, .input_block_col_right {
    width: 50%;
    display: flex;
    justify-content: flex-start;
    flex-direction: column;
    padding-top: 48px;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
}

.input_block_col_left {
    padding-left: 32px;
}

.input_block_col_right {
    align-items: flex-end;
    padding-right: 32px;
}

.input_label {
    width: 100%;
    height: 100%;
    position: relative;
    text-align: right;
}

.input_block_name, .input_prefix {
    font-size: 36px;
}

.input_block_balance, .input_block_bottom_label {
    font-size: 12px;
    padding-top: 12px;
}

.input_block_balance {
    opacity: 0.4;
}

.input_block_currency_symbol {
    font-size: 10px;
}
