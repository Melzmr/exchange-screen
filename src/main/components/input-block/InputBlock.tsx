import React from 'react';
import './InputBlock.css'
import {IPocket} from '../../core/reducers/reducePockets';
import {formatNumber} from '../../core/calculation-utils';

export interface IInputBlockProps {
  pocket: IPocket;
  onAmountChange: (pocket: IPocket | '') => void;
  amount: number | '';
  error?: string;
  amountPrefix?: string;
  className?: string;
  bottomLabel?: React.ReactNode;
}

export const InputBlock = ({pocket, onAmountChange, amount, error, amountPrefix, className, bottomLabel}: IInputBlockProps) => {

  const handleOnChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const amount = parseFloat(event.target.value);
    if (!isNaN(amount)) {
      onAmountChange(
          {
            ...pocket,
            value: formatNumber(amount),
          }
      );
    } else {
      onAmountChange('');
    }
  };

  const calculateWidth = () => {
    const amountLength = amount.toString().length;
    if (amountLength) {
      return amountLength + 'ch';
    }
    return '1ch';
  };

  return (
      <div className="input_block_inner_container">
        <div className="input_block_col_left">
          <div className="input_block_name">
            {pocket.name}
          </div>
          <div className="input_block_balance">
            You have <span className="input_block_currency_symbol">{pocket.currency}</span>{pocket.value}
          </div>
        </div>
        <div className="input_block_col_right">
          <label className="input_label">
            {amount && (
                <span className="input_prefix">
                  {amountPrefix}
                </span>
            )}
            <input
                type="number"
                step="0.01"
                style={{width: calculateWidth()}}
                value={amount}
                onChange={handleOnChange}
                inputMode="numeric"
                pattern="[0-9]*"
                className="input_block_amount"
            />
            {bottomLabel && (
                <div className="input_block_bottom_label">
                  {bottomLabel}
                </div>
            )}
          </label>
        </div>
      </div>
  )
};

InputBlock.displayName = 'InputBlock';
