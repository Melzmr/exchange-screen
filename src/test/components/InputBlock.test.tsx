import React from 'react';
import {render} from '@testing-library/react';
import {shallow} from 'enzyme';
import {InputBlock} from '../../main/components/input-block/InputBlock';

describe('InputBlock', () => {
  let comp: React.ReactElement;
  let fn;

  beforeEach(() => {
    fn = jest.fn();
    comp = (
        <InputBlock
            amount={'100'}
            onAmountChange={fn}
            amountPrefix={'+'}
            className={'foo'}
            pocket={{
              name: 'USD',
              value: '100',
              currency: '$',
            }}
            bottomLabel={(
                <span>
                  TestLabel
                </span>
            )}
        />
    );
  });

  it('renders InputBlock component', () => {
    const wrapper = shallow(comp);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders bottom label', () => {
    const {getByText} = render(comp);
    const el = getByText('TestLabel');
    expect(el).toBeInTheDocument();
  });

  it('renders currency of current pocket', () => {
    const {getByText} = render(comp);
    const el = getByText('USD');
    expect(el).toBeInTheDocument();
  });

  it('renders prefix', () => {
    const {getByText} = render(comp);
    const el = getByText('+');
    expect(el).toBeInTheDocument();
  });

  it('adds classname to container', () => {
    const {container} = render(comp);
    expect((container.firstChild as any).classList.contains('foo')).toBeTruthy();
  });

});


