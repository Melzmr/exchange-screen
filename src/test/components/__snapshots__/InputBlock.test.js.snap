// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputBlock renders InputBlock component 1`] = `
<div
  className="input_block_inner_container foo"
>
  <div
    className="input_block_col_left"
  >
    <div
      className="input_block_name"
    >
      USD
    </div>
    <div
      className="input_block_balance"
    >
      You have 
      <span
        className="input_block_currency_symbol"
      >
        $
      </span>
      100
    </div>
  </div>
  <div
    className="input_block_col_right"
  >
    <label
      className="input_label"
    >
      <span
        className="input_prefix"
      >
        +
      </span>
      <Input
        onChange={[Function]}
        style={
          Object {
            "width": "1ch",
          }
        }
        value="100"
      />
      <div
        className="input_block_bottom_label"
      >
        <span>
          TestLabel
        </span>
      </div>
    </label>
  </div>
</div>
`;
